src/instructlab/model/evaluate.py

```diff
@@ -272,85 +224,107 @@ def get_model_name(model_path):
 
 
 def get_cpu_count():
+    """Returns the available cpu count to this process"""
     try:
         # Not available on all platforms
         return len(os.sched_getaffinity(0))  # type: ignore[attr-defined]
     except AttributeError:
         return multiprocessing.cpu_count()
 
 
+def get_gpus(ctx, gpus=None) -> tuple[int | None, int]:
+    """Return the number of gpus explicitly selected through --gpus or config
+    The second value in the tuple is the effective gpus that will be used by
+    serving. If gpus is specified, the two values will be the same. 1 is the min
+    value for effective_gpus.
+    """
+    # First Party
+    from instructlab.model.backends.vllm import get_argument
+
+    eval_serve = ctx.obj.config.serve
+    gpus = gpus or eval_serve.vllm.gpus
+
+    effective_gpus = gpus
+    if effective_gpus is None:
+        try:
+            tps = get_argument("--tensor-parallel-size", eval_serve.vllm.vllm_args)
+            if tps is not None:
+                effective_gpus = int(tps)
+        except ValueError:
+            logger.warning("Invalid --tensor-parallel-size found in serve vllm_args")
+    effective_gpus = effective_gpus or 1
+    return gpus, effective_gpus
+
+
+def get_backend(backend, model):
+    """Return the backend based on specified backend and model detection"""
+    if backend is None:
+        try:
+            return backends.get(pathlib.Path(model), backend)
+        except ValueError as e:
+            click.secho(f"Failed to determine backend: {e}", fg="red")
+            raise click.exceptions.Exit(1)
+    return backend
+
+
 def launch_server(
     ctx: click.Context,
     model: str,
     model_name: str,
-    max_workers: int,
+    max_workers: str | int,
     gpus: int | None,
     backend: str | None,
     enable_serving_output: bool,
 ) -> tuple:
     eval_serve = deepcopy(ctx.obj.config.serve)
-    if backend is None:
-        try:
-            backend = eval_serve.backend = backends.get(pathlib.Path(model), backend)
-        except ValueError as e:
-            click.secho(f"Failed to determine backend: {e}", fg="red")
-            raise click.exceptions.Exit(1)
+    eval_serve.backend = backend = get_backend(backend, model)
 
+    effective_gpus = 1
```

Comment:
@danmcp What happens if the case when we want to run with llama-cpp and there are no gpus and what should the intended behavior be? 

From tracing through the code, I notice that with this assignment, `launch_server()`  would return `effective_gpus = 1` if the backend is LLAMA_CPP. Then this value would go into `evaluator.gen_answers(serving_gpus=effective_gpus)`. 

Digging into the eval library code `mt_bench_answers.generate_answers()` takes in `serving_gpus` when trying to calculate the max workers. `AbstractMTBenchEvaluator.get_effective_max_workers()` takes `serving_gpus` and calls `self._calc_max_workers()` and there it checks if `serving_gpus is not None`. The `else` in that conditional calculates the max workers using a cpu count (https://github.com/instructlab/eval/blob/main/src/instructlab/eval/mt_bench.py#L79).

 In the cpu only case with llama-cpp would we want to pass `0` for `effective_gpus` all the way down or `None`?

Reply:
> In the cpu only case with llama-cpp would we want to pass 0 for effective_gpus all the way down or None?

1 serves as a min for the purposes of tuning max_workers.

So for the purposes of interacting with the eval library:
Passing None falls into catch-all code block where we haven't been provided any info on the number of gpus.  We don't want that since we do know something about the number of gpus being used in this case.  Passing 1 or 0 would have the same effect since 1 is the min for tuning purposes.

For the purposes of the rest of the code in evaluate.py.  1 needs to be the min for the reset of the logic to work.  I would add another max(1, effective_gpus) below and use 0 here.

I was guessing this might be a bit confusing.  Since we are only using effective_gpus for tuning another param, the value didn't need to be accurate for 1 vs 0.  But it's possible that logic could be misunderstood or misused later on, so let me see if I can make some adjustments to remove the chance of that confusion

Reply:
Updated, let me know what you think!